src/Controller/BulkController.php

```diff
@@ -0,0 +1,249 @@
+            $this->categories = (array)Hash::extract($response, 'data.{n}.attributes.name');
+        } catch (BEditaClientException $ex) {
+            $this->errors[] = ['message' => $ex->getAttributes()];
+        }
+    }
+
+    /**
+     * Save categories for selected objects.
+     *
+     * @return void
+     */
+    protected function saveCategories(): void
+    {
+        foreach ($this->ids as $id) {
+            try {
+                $object = $this->apiClient->getObject($id, $this->objectType);
+                $objectCategories = (array)Hash::extract($object, 'data.attributes.categories.{n}.name');
+                $this->categories = !empty($objectCategories) ? array_unique(array_merge((array)$this->categories, $objectCategories)) : $this->categories;
+                $payload = compact('id');
+                $payload['categories'] = $this->remapCategories((array)$this->categories);
+                $this->apiClient->save($this->objectType, $payload);
+            } catch (BEditaClientException $e) {
+                $this->errors[] = ['id' => $id, 'message' => $e->getAttributes()];
+            }
+        }
+    }
+
+    /**
+     * Remap categories, returning an array of items 'name':<category>
+     *
+     * @param array $input The input categories
+     * @return array
+     */
+    protected function remapCategories(array $input): array
+    {
+        return array_map(function ($category) {
+
+            return ['name' => $category];
+        }, $input);
+    }
+
+    /**
+     * Copy selected objects to position.
+     *
+     * @param string $position The folder ID
+     * @return void
+     */
+    protected function copyToPosition(string $position): void
+    {
+        $payload = [];
+        foreach ($this->ids as $id) {
+            $payload[] = compact('id') + ['type' => $this->objectType];
+        }
+        try {
+            $this->apiClient->addRelated($position, 'folders', 'children', $payload);
+        } catch (BEditaClientException $e) {
+            $this->errors[] = ['message' => $e->getAttributes()];
+        }
+    }
+
+    /**
+     * Move selected objects to position.
+     *
+     * @param string $position The folder ID
+     * @return void
+     */
+    protected function moveToPosition(string $position): void
+    {
+        $payload = ['id' => $position, 'type' => 'folders'];
+        foreach ($this->ids as $id) {
+            try {
+                $this->apiClient->replaceRelated($id, $this->objectType, 'parents', $payload);
+            } catch (BEditaClientException $e) {
+                $this->errors[] = ['id' => $id, 'message' => $e->getAttributes()];
+            }
+        }
+    }
+
+    /**
+     * Handle page errors.
+     *
+     * @return void
+     */
+    protected function errors(): void
+    {
+        if (empty($this->errors)) {
+            return;
+        }
+        $this->log($this->errors, LogLevel::ERROR);
```

Comment:
Not sure if  `$this->log()` with array as message works as expected... :thinking: